enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,107 @@
+---
+title: upgrade-preflight-checks-for-operators
+authors:
+  - "@LalatenduMohanty"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-06-04
+last-updated: 2020-06-05
+status: implementable
+---
+
+# Upgrade Preflight Checks For Operators
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+* How CVO can get the list of operators which has the preflight checks implemented?
+* What information CVO needs from the operators to create the preflight jobs?
+* Does CVO already has capability to create jobs for preflight checks?
+* How to trigger execution of preflight checks?
+* How CVO will report when it gets the result from preflight checks?
+* What information CVO needs from operators to create the preflight job?
+* How will preflight checks report back success or failure, and the detailed info for administrators?
+
+## Summary
+
+The goal of over the air update(OTA) feature of OpenShift is to make upgrades seamless and make the upgrade a success almost always.
+To improve the upgrade success we should be able to detect if the new version is compatible with the existing version before running the actual upgrade.
+As OpenShift V4 is composed of many operators, we need to give operators ability to run some checks to find out if the new version is compatible with the existing version and use the result to improve the success rate of upgrades.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before the actual upgrade.
+It would help operators to find out if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered to run these preflight checks without doing an upgrade.
+
+
+### Goals
+
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
+* Preflight checks would be idempotent in nature.
```

Comment:
Is enforcing this a goal?  For example, do we work out a way to run preflights with RBAC that gives them read access but not create/edit/delete access?  Or is that just a nice-to-have refinement that we don't consider to be blocking?

Reply:
Running with RBAC that lets you read everything in the cluster but modify nothing might be a logical starting point to keep things in check.

Reply:
If the preflights has to be idempotent otherwise we are missing the point of this.

Reply:
> If the preflights has to be idempotent otherwise we are missing the point of this.

We all agree on that.  But current goals do not say whether the implementation will _force_ that (e.g. via read-only RBAC), or will give preflights write-enabled RBAC and just trust preflight authors to not cheat.

And even with read-only RBAC, I guess there's a "read platform secret -> create a zillion instances" loophole.